File: Set/src/AVLTree.java
import java.util.ArrayList;

/**
 * 使用映射版本的二元樹來建構AVL樹
 */

public class AVLTree<K extends Comparable<K>, V> {

    private class Node {
        public K key;
        public V value;
        public Node left, right;
        public int height; // 樹高

        public Node(K key, V value) {
            this.key = key;
            this.value = value;
            this.right = null;
            this.left = null;
            this.height = 1;
        }

        public Node(K key) {
            this(key, null);
        }

        public Node() {
            this(null, null);
        }

        @Override
        public String toString() {
            return this.key.toString() + " : " + this.value.toString();
        }

    }

    private Node root;
    private int size;

    public AVLTree() {
        root = null;
        size = 0;
    }

    public int getSize() {
        return size;
    }

    public boolean isEmpty() {
        return size == 0;
    }

    /**
     * 檢查是否為一棵二元樹
     *
     * @return
     */
    public boolean isBST() {
        if (root == null) {
            return true;
        }

        ArrayList<K> arr = new ArrayList<>();
        inOrder(root, arr);

        for (int i = 0; i < arr.size() - 1; i++) {
            if (arr.get(i).compareTo(arr.get(i + 1)) > 0) {
                return false;
            }
        }

        return true;
    }

    /**
     * 中序走訪
     *
     * @param node
     * @param arr
     */
    private void inOrder(Node node, ArrayList<K> arr) {
        if (node == null) {
            return;
        }

        inOrder(node.left, arr);
        arr.add(node.key);
        inOrder(node.right, arr);
    }

    public boolean isBalanced() {
        return isBalanced(root);
    }

    private boolean isBalanced(Node node) {
        if (node == null) {
            return true;
        }

        if (Math.abs(getBalanceFactor(node)) > 1) {
            return false;
        }

        return isBalanced(node.left) && isBalanced(node.right);
    }


    /**
     * 獲取高度
     */
    private int getHeight(Node node) {
        if (node == null) {
            return 0;
        }
        return node.height;
    }

    public void add(K key, V value) {
        root = add(root, key, value);
    }

    public V remove(K key) {
        root = remove(root, key);
        return null;
    }

    public boolean contains(K key) {
        return getNode(root, key) != null;
    }

    public V get(K key) {
        Node node = getNode(root, key);
        return (node == null) ? null : node.value;
    }

    public void set(K key, V newValue) {
        Node node = getNode(root, key);
        if (node == null) {
            throw new IllegalArgumentException("key : " + key + " doesn't exist!");
        }

        node.value = newValue;
    }

    private int getBalanceFactor(Node node) {
        if (node == null) {
            return 0;
        }

        return getHeight(node.left) - getHeight(node.right);
    }

    // 對節點y進行向右旋轉操作，返回旋轉後新的根節點x
    //        y                              x
    //       / \                           /   \
    //      x   T4     向右旋轉 (y)        z     y
    //     / \       - - - - - - - ->    / \   / \
    //    z   T3                       T1  T2 T3 T4
    //   / \
    // T1   T2
    private Node rightRotate(Node y) {
        Node x = y.left;
        Node T3 = x.right;

        // 右旋轉
        x.right = y;
        y.left = T3;

        // 更新height
        y.height = 1 + Math.max(getHeight(y.left), getHeight(y.right));
        x.height = 1 + Math.max(getHeight(x.left), getHeight(x.right));

        return x;
    }

    // 對節點y進行向左旋轉操作，返回旋轉後新的根節點x
    //    y                             x
    //  /  \                          /   \
    // T1   x      向左旋轉 (y)       y     z
    //     / \   - - - - - - - ->   / \   / \
    //   T2  z                     T1 T2 T3 T4
    //      / \
    //     T3 T4
    private Node leftRotate(Node y) {
        Node x = y.right;
        Node T2 = x.left;

        // 左旋轉
        x.left = y;
        y.right = T2;

        // 更新height
        y.height = 1 + Math.max(getHeight(y.left), getHeight(y.right));
        x.height = 1 + Math.max(getHeight(x.left), getHeight(x.right));

        return x;
    }

    /**
     * 插入節點並保持樹平衡
     *
     * @param node
     * @param key
     * @param value
     * @return
     */
    private Node add(Node node, K key, V value) {
        if (node == null) {
            size++;
            return new Node(key, value);
        }

        int originalHeight = getHeight(node);

        if (key.compareTo(node.key) < 0) {
            node.left = add(node.left, key, value);
        } else if (key.compareTo(node.key) > 0) {
            node.right = add(node.right, key, value);
        } else {
            node.value = value;
        }

        node.height = 1 + Math.max(getHeight(node.left), getHeight(node.right));

        // 高度相同則不需要改變
        if (originalHeight == node.height) {
            return node;
        }

        // 計算平衡因子
        int balanceFactor = getBalanceFactor(node);

        // LL
        if (balanceFactor > 1 && getBalanceFactor(node.left) >= 0) {
            return rightRotate(node);
        }

        // RR
        if (balanceFactor < -1 && getBalanceFactor(node.right) <= 0) {
            return leftRotate(node);
        }

        // LR
        if (balanceFactor > 1 && getBalanceFactor(node.left) < 0) {
            node.left = leftRotate(node.left);
            return rightRotate(node);
        }

        // RL
        if (balanceFactor < -1 && getBalanceFactor(node.right) > 0) {
            node.right = rightRotate(node.right);
            return leftRotate(node);
        }

        return node;
    }

    private Node getNode(Node node, K key) {
        if (node == null) {
            return null;
        }

        if (key.compareTo(node.key) == 0) {
            return node;
        } else if (key.compareTo(node.key) < 0) {
            return getNode(node.left, key);
        } else {
            return getNode(node.right, key);
        }
    }

    private Node min(Node node) {
        if (node.left == null) {
            return node;
        }
        return min(node.left);
    }

    /**
     * 刪除節點並保持平衡
     *
     * @param node
     * @param key
     * @return
     */
    private Node remove(Node node, K key) {
        if (node == null) {
            return null;
        }

        Node retNode;
        if (key.compareTo(node.key) > 0) {
            node.right = remove(node.right, key);
            retNode = node;
        } else if (key.compareTo(node.key) < 0) {
            node.left = remove(node.left, key);
            retNode = node;
        } else {
            if (node.left == null) {
                Node rightNode = node.right;
                node.right = null;
                size--;
                retNode = rightNode;
            } else if (node.right == null) {
                Node leftNode = node.left;
                node.left = null;
                size--;
                retNode = leftNode;
            } else {
                Node successor = min(node.right);
                successor.right = remove(node.right, successor.key);
                successor.left = node.left;
                node.left = node.right = null;
                retNode = successor;
            }
        }

        if (retNode == null) {
            return null;
        }

        retNode.height = 1 + Math.max(getHeight(retNode.left), getHeight(retNode.right));

        int balanceFactor = getBalanceFactor(retNode);

        if (balanceFactor > 1 && getBalanceFactor(retNode.left) >= 0) {
            return rightRotate(retNode);
        }

        if (balanceFactor < -1 && getBalanceFactor(retNode.right) <= 0) {
            return leftRotate(retNode);
        }

        if (balanceFactor > 1 && getBalanceFactor(retNode.left) < 0) {
            retNode.left = leftRotate(retNode.left);
            return rightRotate(retNode);
        }

        if (balanceFactor < -1 && getBalanceFactor(retNode.right) > 0) {
            retNode.right = rightRotate(retNode.right);
            return leftRotate(retNode);
        }

        return retNode;
    }
}

File: C _version/src/avltree.c
#include "avltree.h"

static __inline int max(int a, int b){

    return a > b ? a : b;
}

static __inline int getHeight(Node* node){

    if(node == NULL){
        return 0;
    }
    return node->height;
}

static __inline int getBalanceFactor(Node* node){

    if(node == NULL){
        return 0;
    }
    return getHeight(node->left) - getHeight(node->right);
}

static Node* newNode(int value){

    Node* node = (Node*)malloc(sizeof(Node));
    node->value = value;
    node->left = NULL;
    node->right = NULL;
    node->height = 1;

    return node;
}

static Node* rightRotate(Node* y){

    Node* x = y->left;
    Node* T3 = x->right;

    x->right = y;
    y->left = T3;

    y->height = max(getHeight(y->left),getHeight(y->right)) + 1;
    x->height = max(getHeight(x->left),getHeight(x->right)) + 1;

    return x;
}

static Node* leftRotate(Node* y){

    Node* x = y->right;
    Node* T2 = x->left;

    x->left = y;
    y->right = T2;

    y->height = max(getHeight(y->left),getHeight(y->right)) + 1;
    x->height = max(getHeight(x->left),getHeight(x->right)) + 1;

    return x;
}

static Node* add(Node* node, int value, int* size){

    if(node == NULL){
        ++(*size);
        return newNode(value);
    }
    if(value < node->value){
        node->left = add(node->left,value,size);
    }else if(value > node->value){
        node->right = add(node->right,value,size);
    }

    node->height = 1 + max(getHeight(node->left),getHeight(node->right));

    int balanceFactor = getBalanceFactor(node);

    if(balanceFactor > 1 && (value < node->left->value)){
        return rightRotate(node);
    }
    if(balanceFactor < -1 && (value > node->right->value)){
        return leftRotate(node);
    }
    if(balanceFactor > 1 && (value > node->left->value)){
        node->left = leftRotate(node->left);
        return rightRotate(node);
    }
    if(balanceFactor < -1 && (value < node->right->value)){
        node->right = rightRotate(node->right);
        return leftRotate(node);
    }

    return node;
}

void avl_insert(AVLTree* tree, int value){

    tree->root = add(tree->root,value,&(tree->size));
}

static Node* minValueNode(Node* node){

    Node* current = node;
    while(current->left != NULL){
        current = current->left;
    }

    return current;
}

static Node* removeNode(Node* root, int value, int* size){

    if(root == NULL){
        return root;
    }
    if(value < root->value){
        root->left = removeNode(root->left,value,size);
    }else if(value > root->value){
        root->right = removeNode(root->right,value,size);
    }else{
        (*size)--;
        if((root->left == NULL) || (root->right == NULL)){
            Node* temp = root->left ? root->left : root->right;
            if(temp == NULL){
                temp =  root;
                root = NULL;
            }else{
                *root = *temp;
            }
            free(temp);
        }else{
            Node* temp = minValueNode(root->right);
            root->value = temp->value;
            root->right = removeNode(root->right,temp->value,size);
        }
    }
    if(root == NULL){ 
        return root;
    }

    root->height = 1 + max(getHeight(root->left),getHeight(root->right));

    int balance = getBalanceFactor(root);

    if(balance > 1 && getBalanceFactor(root->left) >= 0){ 
        return rightRotate(root);
    }
    if(balance > 1 && getBalanceFactor(root->left) < 0){
        root->left = leftRotate(root->left);
        return rightRotate(root);
    }
    if(balance < -1 && getBalanceFactor(root->right) <= 0){
        return leftRotate(root);
    }
    if(balance < -1 && getBalanceFactor(root->right) > 0){
        root->right = rightRotate(root->right);
        return leftRotate(root);
    }

    return root;
}

void avl_remove(AVLTree* tree, int value){

    tree->root = removeNode(tree->root,value,&(tree->size));
}

static bool isBalanced(Node* node){

    if(node == NULL){
        return true;
    }
    if(abs(getBalanceFactor(node)) > 1){
        return false;
    }

    return isBalanced(node->left) &&  isBalanced(node->right);
}

bool avl_isBalanced(AVLTree* tree){

    return isBalanced(tree->root);
}

static bool isBST(Node* node, int max, int min){

    if(node == NULL){
        return true;
    }
    if(node->value < min || node->value > max){
        return false;
    }

    return isBST(node->left,node->value,min) && isBST(node->right,max,node->value);
}

bool avl_isBST(AVLTree* tree){

    return isBST(tree->root,INT_MAX,INT_MIN);
}

AVLTree* avl_createTree(){

    AVLTree* tree = (AVLTree*)malloc(sizeof(AVLTree));
    tree->root = NULL;
    tree->size = 0;
    return tree;
}
